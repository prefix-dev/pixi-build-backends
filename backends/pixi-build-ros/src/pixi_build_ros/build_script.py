"""
Build script generation for Python backend.
"""

from enum import Enum
from pathlib import Path
from typing import Any, Dict, List
import platform


class Installer(Enum):
    """Available Python installers."""

    UV = "uv"
    PIP = "pip"

    def package_name(self) -> str:
        """Get the package name for this installer."""
        return self.value

    @classmethod
    def determine_installer(cls, dependencies: Dict[str, Any]) -> "Installer":
        """Determine which installer to use based on dependencies."""
        if "uv" in dependencies:
            return cls.UV
        return cls.PIP


class BuildPlatform(Enum):
    """Build platform types."""

    WINDOWS = "windows"
    UNIX = "unix"

    @classmethod
    def current(cls) -> "BuildPlatform":
        """Get current build platform."""
        return cls.WINDOWS if platform.system() == "Windows" else cls.UNIX


class BuildScriptContext:
    """Context for build script generation."""

    def __init__(
        self,
        installer: Installer,
        build_platform: BuildPlatform,
        editable: bool,
        manifest_root: Path,
    ):
        self.installer = installer
        self.build_platform = build_platform
        self.editable = editable
        self.manifest_root = manifest_root

    def render(self) -> List[str]:
        """Render the build script."""
        if self.build_platform == BuildPlatform.WINDOWS:
            python_var = "%PYTHON%"
            src_dir = str(self.manifest_root) if self.editable else "%SRC_DIR%"
        else:
            python_var = "$PYTHON"
            src_dir = str(self.manifest_root) if self.editable else "$SRC_DIR"

        editable_option = " --editable" if self.editable else ""
        common_options = f"-vv --no-deps --no-build-isolation{editable_option}"

        if self.installer == Installer.UV:
            command = f"uv pip install --python {python_var} {common_options} {src_dir}"
        else:
            command = f"{python_var} -m pip install --ignore-installed {common_options} {src_dir}"

        lines = [command]

        if self.build_platform == BuildPlatform.WINDOWS:
            lines.append("if errorlevel 1 exit 1")

        return lines
