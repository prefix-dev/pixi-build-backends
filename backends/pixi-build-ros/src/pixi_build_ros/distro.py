from rosdistro import get_cached_distribution, get_index, get_index_url


class Distro(object):
    def __init__(self, distro_name):
        index = get_index(get_index_url())
        self._distro = get_cached_distribution(index, distro_name)
        self.distro_name = distro_name

        # cache distribution type
        self._distribution_type = index.distributions[distro_name]["distribution_type"]
        self._python_version = index.distributions[distro_name]["python_version"]

    @property
    def name(self) -> str:
        return self.distro_name

    def check_ros1(self):
        return self._distribution_type == "ros1"

    def get_python_version(self):
        return self._python_version

    def get_package_names(self):
        return self._distro.release_packages.keys()

    def has_package(self, package_name):
        """Check if the distribution has a specific package."""
        return package_name in self._distro.release_packages
