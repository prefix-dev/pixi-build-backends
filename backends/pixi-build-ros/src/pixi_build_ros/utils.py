import os
from pathlib import Path
from typing import Any, Optional, Dict, List
import re
import toml


def parse_entry_points_from_scripts(scripts: Dict[str, str]) -> List[str]:
    """
    Parse entry points from a dictionary of scripts.

    This function validates entry point format and converts a scripts dictionary
    to a list of properly formatted entry point strings.

    Parameters
    ----------
    scripts : Dict[str, str]
        Dictionary mapping script names to entry points (e.g., {"script": "module:function"})

    Returns
    -------
    List[str]
        List of formatted entry point strings (e.g., ["script = module:function"])

    Raises
    ------
    ValueError
        If any entry point has invalid format

    Examples
    --------
    >>> scripts = {"my_script": "mymodule:main", "other": "pkg.mod:func"}
    >>> parse_entry_points_from_scripts(scripts)
    ['my_script = mymodule:main', 'other = pkg.mod:func']
    """
    if not scripts:
        return []

    entry_points = []
    # Entry point format validation regex: name should be valid Python identifier-like,
    # and entry point should be in format "module:function" or "module.submodule:function"
    entry_point_pattern = re.compile(r"^[a-zA-Z_][a-zA-Z0-9_]*(\.[a-zA-Z_][a-zA-Z0-9_]*)*:[a-zA-Z_][a-zA-Z0-9_]*$")

    for name, entry_point in scripts.items():
        # Validate entry point format
        if not entry_point_pattern.match(entry_point):
            raise ValueError(f"Invalid entry point format: {entry_point}")

        # Format as "name = entry_point"
        formatted_entry_point = f"{name} = {entry_point}"
        entry_points.append(formatted_entry_point)

    return entry_points


def extract_entry_points(pyproject_manifest: Optional[Dict[str, Any]]) -> List[str]:
    """
    Extract entry points from pyproject.toml.

    Parameters
    ----------
    pyproject_manifest : Optional[dict]
        The pyproject.toml manifest dictionary.

    Returns
    -------
    list
        A list of entry points, or empty list if no scripts found.

    Examples
    --------
    ```python
    >>> manifest = {"project": {"scripts": {"my_script": "module:function"}}}
    >>> extract_entry_points(manifest)
    ['my_script = module:function']
    >>> extract_entry_points(None)
    []
    >>> extract_entry_points({})
    []
    >>>
    ```
    """
    if not pyproject_manifest:
        return []

    project = pyproject_manifest.get("project", {})
    scripts: Dict[str, str] = project.get("scripts", {})

    if not scripts:
        return []

    # Use the Python implementation instead of FFI
    return parse_entry_points_from_scripts(scripts)


def read_pyproject_toml(manifest_root: Path) -> Optional[Dict[str, Any]]:
    """Read pyproject.toml if it exists."""
    pyproject_path = manifest_root / "pyproject.toml"
    if pyproject_path.exists():
        return toml.load(pyproject_path)
    return None


def get_build_input_globs(config: Any, workdir: Path, editable: bool) -> List[str]:
    """Get build input globs for Python package."""
    base_globs = [
        # Source files
        "**/*.c",
        "**/*.cpp",
        "**/*.rs",
        "**/*.sh",
        # Common data files
        "**/*.json",
        "**/*.yaml",
        "**/*.yml",
        "**/*.txt",
        # Project configuration
        "setup.py",
        "setup.cfg",
        "pyproject.toml",
        "requirements*.txt",
        "Pipfile",
        "Pipfile.lock",
        "poetry.lock",
        "tox.ini",
        # Build configuration
        "Makefile",
        "MANIFEST.in",
        "tests/**/*.py",
        "docs/**/*.rst",
        "docs/**/*.md",
        # Versioning
        "VERSION",
        "version.py",
    ]

    python_globs = [] if editable else ["**/*.py", "**/*.pyx"]

    all_globs = base_globs + python_globs
    if hasattr(config, "extra_input_globs"):
        all_globs.extend(config.extra_input_globs)

    return all_globs


def get_editable_setting(python_params: Any) -> bool:
    """Get editable setting from environment or params."""
    env_editable = os.environ.get("BUILD_EDITABLE_PYTHON", "").lower() == "true"
    if env_editable:
        return True

    if python_params and hasattr(python_params, "editable"):
        return bool(python_params.editable)

    return False
