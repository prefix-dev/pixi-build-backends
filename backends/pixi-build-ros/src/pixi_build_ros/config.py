import os
import re
from pathlib import Path
from typing import Any

import pydantic
import yaml

from pixi_build_ros.distro import Distro


def _parse_str_as_abs_path(value: str | Path, manifest_root: Path) -> Path:
    """Parse a string as a Path."""
    # Ensure the debug directory is a Path object
    if isinstance(value, str):
        value = Path(value)
    # Ensure it's an absolute path
    if not value.is_absolute():
        # Convert to absolute path relative to manifest root
        return (manifest_root / value).resolve()
    return value


def _extract_distro_from_channels_list(channels: list[str] | None) -> str | None:
    """Extract ROS distro from a list of channel URLs/names.

    Looks for channels matching the pattern 'robostack-<distro>' and returns
    the first distro found.

    Args:
        channels: List of channel URLs or names

    Returns:
        The distro name if found, None otherwise
    """
    if not channels:
        return None

    # Pattern to match robostack-<distro> in channel URLs or names
    robostack_pattern = re.compile(r"robostack-(\w+)")

    for channel in channels:
        # Extract the last path component or the whole channel name if it's not a URL
        # This handles both "robostack-humble" and "https://prefix.dev/robostack-humble"
        channel_name = channel.rstrip("/").split("/")[-1]
        match = robostack_pattern.search(channel_name)
        if match:
            return match.group(1)

    return None


PackageMapEntry = dict[str, list[str] | dict[str, list[str]]]


class PackageMappingSource:
    """Describes where additional package mapping data comes from."""

    def __init__(self, mapping: dict[str, PackageMapEntry], source_file: Path | None = None):
        if mapping is None:
            raise ValueError("PackageMappingSource mapping cannot be null.")
        if not isinstance(mapping, dict):
            raise TypeError("PackageMappingSource mapping must be a dictionary.")
        # Copy to keep the source immutable for callers.
        self.mapping: dict[str, PackageMapEntry] = dict(mapping)
        # Track the source file path if this came from a file
        self.source_file: Path | None = source_file

    @classmethod
    def from_mapping(cls, mapping: dict[str, PackageMapEntry]) -> "PackageMappingSource":
        """Create a source directly from a mapping dictionary."""
        return cls(mapping, source_file=None)

    @classmethod
    def from_file(cls, file_path: str | Path) -> "PackageMappingSource":
        """Create a source from a mapping file."""
        path = Path(file_path)
        if not path.exists():
            raise ValueError(f"Additional package map file '{path}' not found.")
        with open(path) as f:
            data = yaml.safe_load(f) or {}
        if not isinstance(data, dict):
            raise TypeError("Expected package map file to contain a dictionary.")
        return cls(data, source_file=path)

    def get_package_mapping(self) -> dict[str, PackageMapEntry]:
        return dict(self.mapping)

    def get_source_file(self) -> Path | None:
        """Return the source file path if this mapping came from a file."""
        return self.source_file


class ROSBackendConfig(pydantic.BaseModel, extra="forbid", arbitrary_types_allowed=True):
    """ROS backend configuration."""

    # ROS distribution to use, e.g., "foxy", "galactic", "humble"
    # Can be auto-detected from robostack- channel if not explicitly specified
    distro: Distro | None = None

    noarch: bool | None = None
    # Environment variables to set during the build
    env: dict[str, str] | None = None
    # Directory for debug files of this script
    debug_dir: Path | None = pydantic.Field(default=None, alias="debug-dir")
    # Extra input globs to include in the build hash
    extra_input_globs: list[str] | None = pydantic.Field(default=None, alias="extra-input-globs")

    # Extra package mappings to use in the build
    extra_package_mappings: list[PackageMappingSource] = pydantic.Field(
        default_factory=list, alias="extra-package-mappings"
    )

    def is_noarch(self) -> bool:
        """Whether to build a noarch package or a platform-specific package."""
        return self.noarch is None or self.noarch

    def get_package_mapping_file_paths(self) -> list[Path]:
        """Get all file paths from package mappings that came from files."""
        file_paths = []
        for source in self.extra_package_mappings:
            if source_file := source.get_source_file():
                file_paths.append(source_file)
        return file_paths

    @pydantic.field_validator("distro", mode="before")
    @classmethod
    def _parse_distro(cls, value: Any) -> Distro | None:
        """Parse a distro string."""
        if isinstance(value, str):
            return Distro(value)
        if isinstance(value, Distro):
            return value
        return None

    @classmethod
    def resolve_distro(
        cls,
        config: "ROSBackendConfig",
        channels: list[str] | None = None,
    ) -> "ROSBackendConfig":
        """Resolve the distro field, auto-detecting from channels if not explicitly set.

        This should be called after config validation to fill in the distro if needed.

        Args:
            config: The config instance (possibly with distro=None)
            channels: List of channel URLs from the build system

        Returns:
            Config with distro resolved

        Raises:
            ValueError: If distro cannot be determined
        """
        # If distro is already set, nothing to do
        if config.distro is not None:
            return config

        # Try to auto-detect from channels
        detected_distro_name = None
        if channels:
            detected_distro_name = _extract_distro_from_channels_list(channels)

        if detected_distro_name:
            config.distro = Distro(detected_distro_name)
            return config

        # If we couldn't detect a distro, raise an error
        raise ValueError(
            "ROS distro must be either explicitly configured or auto-detected from robostack channels. "
            "No robostack channel found in the project."
        )

    @pydantic.model_validator(mode="after")
    def _ensure_distro(
        self,
        info: pydantic.ValidationInfo,
    ) -> "ROSBackendConfig":
        """Ensure distro is resolved after validation."""
        context = info.context or {}
        channels = context.get("channels")
        return self.resolve_distro(self, channels=channels)

    @pydantic.field_validator("debug_dir", mode="before")
    @classmethod
    def _parse_debug_dir(cls, value: Any, info: pydantic.ValidationInfo) -> Path | None:
        """Parse debug directory if set."""
        if value is None:
            return None
        base_path = Path(os.getcwd())
        if info.context and "manifest_root" in info.context:
            base_path = Path(info.context["manifest_root"])
        return _parse_str_as_abs_path(value, base_path)

    @pydantic.field_validator("extra_package_mappings", mode="before")
    @classmethod
    def _parse_package_mappings(
        cls, input_value: Any, info: pydantic.ValidationInfo
    ) -> list[PackageMappingSource] | None:
        """Parse additional package mappings if set."""
        if input_value is None:
            return []

        base_path = Path.cwd()
        if info.context and "manifest_root" in info.context:
            base_path = Path(info.context["manifest_root"])

        result: list[PackageMappingSource] = []
        for raw_entry in input_value:
            # match for cases
            # it's already a package mapping source (usually for testing)
            if isinstance(raw_entry, PackageMappingSource):
                entry = raw_entry
            elif isinstance(raw_entry, dict):
                if "file" in raw_entry:
                    file_value = raw_entry["file"]
                    entry = PackageMappingSource.from_file(_parse_str_as_abs_path(file_value, base_path))
                elif "mapping" in raw_entry:
                    mapping_value = raw_entry["mapping"]
                    entry = PackageMappingSource.from_mapping(mapping_value)
                else:
                    entry = PackageMappingSource.from_mapping(raw_entry)
            elif isinstance(raw_entry, (str, Path)):
                entry = PackageMappingSource.from_file(_parse_str_as_abs_path(raw_entry, base_path))
            else:
                raise ValueError(
                    f"Unrecognized entry for extra-package-mappings: {raw_entry} of type {type(raw_entry)}."
                )
            result.append(entry)
        return result

    def get_distro(self) -> Distro:
        if self.distro is None:
            raise ValueError("Distro could not be resolved from the channels or the `distro` build config.")
        return self.distro
