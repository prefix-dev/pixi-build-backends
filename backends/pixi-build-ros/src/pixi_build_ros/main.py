from .python_generator import PythonGenerator
from pixi_build_backend.main import run_backend


def main() -> None:
    """Main entry point for the script."""
    generator = PythonGenerator()
    run_backend(generator)
