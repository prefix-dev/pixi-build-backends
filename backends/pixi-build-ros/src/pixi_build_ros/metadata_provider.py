"""
ROS-specific metadata provider that extracts metadata from package.xml files.
"""

from typing import Optional, List

from pixi_build_backend.types import MetadataProvider
from pixi_build_ros.distro import Distro


class MaintainerInfo:
    """Container for maintainer information from package.xml."""
    
    def __init__(self, name: str, email: str):
        self.name = name
        self.email = email


class PackageData:
    """Container for parsed package.xml data."""
    
    def __init__(
        self,
        name: Optional[str] = None,
        version: Optional[str] = None,
        description: Optional[str] = None,
        maintainers: Optional[List[MaintainerInfo]] = None,
        licenses: Optional[List[str]] = None,
        homepage: Optional[str] = None,
        repository: Optional[str] = None,
    ):
        self.name = name
        self.version = version
        self.description = description
        self.maintainers = maintainers or []
        self.licenses = licenses or []
        self.homepage = homepage
        self.repository = repository

class PackageXmlMetadataProvider(MetadataProvider):
    """
    Metadata provider that extracts metadata from ROS package.xml files.

    This provider reads ROS package.xml files and extracts package metadata
    like name, version, description, maintainers, etc.
    """

    def __init__(self, package_xml_path: str, *args, **kwargs):
        """
        Initialize the metadata provider with a package.xml file path.

        Args:
            package_xml_path: Path to the package.xml file
        """
        super().__init__(*args, **kwargs)
        self.package_xml_path = package_xml_path
        self._package_data: Optional[PackageData] = None

    @property
    def _package_xml_data(self) -> PackageData:
        """Load and parse the package.xml file."""
        if self._package_data is not None:
            return self._package_data

        try:
            import xml.etree.ElementTree as ET

            tree = ET.parse(self.package_xml_path)
            root = tree.getroot()

            # Extract basic package information
            name_elem = root.find('name')
            version_elem = root.find('version')
            description_elem = root.find('description')

            # Extract maintainer and author information
            maintainers: List[MaintainerInfo] = []
            for maintainer in root.findall('maintainer'):
                maintainer_info = MaintainerInfo(
                    name=maintainer.text.strip() if maintainer.text else '',
                    email=maintainer.get('email', '')
                )
                maintainers.append(maintainer_info)

            # Extract license information
            licenses = []
            for license_elem in root.findall('license'):
                if license_elem.text:
                    licenses.append(license_elem.text.strip())

            # Extract URL information
            homepage = None
            repository = None
            for url in root.findall('url'):
                url_type = url.get('type', '')
                if url_type == 'website' and not homepage:
                    homepage = url.text.strip() if url.text else None
                elif url_type == 'repository' and not repository:
                    repository = url.text.strip() if url.text else None

            self._package_data = PackageData(
                name=name_elem.text.strip() if name_elem is not None and name_elem.text else None,
                version=version_elem.text.strip() if version_elem is not None and version_elem.text else None,
                description=description_elem.text.strip() if description_elem is not None and description_elem.text else None,
                maintainers=maintainers,
                licenses=licenses,
                homepage=homepage,
                repository=repository,
            )

        except Exception as e:
            print(f"Warning: Failed to parse package.xml at {self.package_xml_path}: {e}")
            self._package_data = PackageData()

        return self._package_data

    def name(self) -> Optional[str]:
        """Return the package name from package.xml."""
        return self._package_xml_data.name

    def version(self) -> Optional[str]:
        """Return the package version from package.xml."""
        return self._package_xml_data.version

    def homepage(self) -> Optional[str]:
        """Return the homepage URL from package.xml."""
        return self._package_xml_data.homepage

    def license(self) -> Optional[str]:
        """Return the license from package.xml."""
        # TODO: Handle License parsing to conform to SPDX standards,
        # ROS package.xml does not enforce SPDX as strictly as rattler-build
        return None

    def license_file(self) -> Optional[str]:
        """Return None as package.xml doesn't typically specify license files."""
        return None

    def summary(self) -> Optional[str]:
        """Return the description as summary from package.xml."""
        description = self._package_xml_data.description
        if description and len(description) > 100:
            # Truncate long descriptions for summary
            return description[:97] + "..."
        return description

    def description(self) -> Optional[str]:
        """Return the full description from package.xml."""
        return self._package_xml_data.description

    def documentation(self) -> Optional[str]:
        """Return None as package.xml doesn't typically specify documentation URLs separately."""
        return None

    def repository(self) -> Optional[str]:
        """Return the repository URL from package.xml."""
        return self._package_xml_data.repository

    def input_globs(self) -> List[str]:
        """Return input globs that affect this metadata provider."""
        return [
            "package.xml",
            "CMakeLists.txt", 
            "setup.py",
            "setup.cfg"
        ]


class ROSPackageXmlMetadataProvider(PackageXmlMetadataProvider):
    """
    ROS-specific metadata provider that formats names according to ROS conventions.

    This provider extends PackageXmlMetadataProvider to format package names
    as 'ros-<distro>-<package_name>' according to ROS conda packaging conventions.
    """

    def __init__(self, package_xml_path: str, distro: Optional[Distro] = None):
        """
        Initialize the ROS metadata provider.

        Args:
            package_xml_path: Path to the package.xml file
            distro: ROS distro. If None, will use the base package name without distro prefix.
        """
        super().__init__(package_xml_path)
        self._distro: Optional[Distro] = distro

    def _get_distro(self) -> Optional[Distro]:
        return self._distro

    def name(self) -> Optional[str]:
        """Return the ROS-formatted package name (ros-<distro>-<name>)."""
        base_name = super().name()
        if base_name is None:
            return None

        distro = self._get_distro()
        if distro:
            # Convert underscores to hyphens per ROS conda naming conventions
            formatted_name = base_name.replace('_', '-')
            return f"ros-{distro.name}-{formatted_name}"
        return base_name
