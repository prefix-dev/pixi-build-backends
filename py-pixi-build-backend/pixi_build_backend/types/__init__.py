# Re-export all the types
from .metadata_provider import MetadataProvider as MetadataProvider