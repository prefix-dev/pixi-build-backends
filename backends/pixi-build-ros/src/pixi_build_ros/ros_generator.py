"""
Python generator implementation using Python bindings.
"""

from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Optional, List, Any
from pixi_build_backend.types.generated_recipe import (
    GenerateRecipeProtocol,
    GeneratedRecipe,
)
from .metadata_provider import ROSPackageXmlMetadataProvider
from pixi_build_backend.types.intermediate_recipe import Script, ConditionalRequirements

from pixi_build_backend.types.item import ItemPackageDependency
from pixi_build_backend.types.platform import Platform
from pixi_build_backend.types.project_model import ProjectModelV1
from pixi_build_backend.types.python_params import PythonParams

from .build_script import BuildScriptContext, BuildPlatform
from .distro import Distro
from .utils import get_build_input_globs, package_xml_to_conda_requirements, convert_package_xml_to_catkin_package, \
    get_package_xml_content


@dataclass
class ROSBackendConfig:
    """ROS backend configuration."""

    noarch: Optional[bool] = None
    # Environment variables to set during the build
    env: Optional[Dict[str, str]] = None
    # Directory for debug files of this script
    debug_dir: Optional[Path] = None
    # Extra input globs to include in the build hash
    extra_input_globs: Optional[List[str]] = None
    # ROS distribution to use, e.g., "foxy", "galactic", "humble"
    # TODO: This should be figured out in some other way, not from the config.
    distro: Optional[str] = None

    def is_noarch(self) -> bool:
        """Whether to build a noarch package or a platform-specific package."""
        return self.noarch is None or self.noarch

    def get_debug_dir(self) -> Optional[Path]:
        """Get debug directory if set."""
        if self.debug_dir is not None:
            # Ensure the debug directory is a Path object
            if isinstance(self.debug_dir, str):
                self.debug_dir = Path(self.debug_dir)
            # Ensure it's an absolute path
            if not self.debug_dir.is_absolute():
                # Convert to absolute path relative to the current working directory
                self.debug_dir = Path.cwd() / self.debug_dir
        return self.debug_dir

class ROSGenerator(GenerateRecipeProtocol):
    """ROS recipe generator using Python bindings."""

    def generate_recipe(
        self,
        model: ProjectModelV1,
        config: Dict[str, Any],
        manifest_path: str,
        host_platform: Platform,
        python_params: Optional[PythonParams] = None,
    ) -> GeneratedRecipe:
        """Generate a recipe for a Python package."""
        backend_config: ROSBackendConfig = ROSBackendConfig(**config)

        manifest_root = Path(manifest_path)

        # Setup ROS distro first
        distro = Distro(backend_config.distro)
        
        # Create metadata provider for package.xml
        package_xml_path = manifest_root / "package.xml"
        metadata_provider = ROSPackageXmlMetadataProvider(str(package_xml_path), distro)

        # Create base recipe from model with metadata provider
        generated_recipe = GeneratedRecipe.from_model(model, metadata_provider)

        # Read package.xml for dependency extraction
        package_xml_str = get_package_xml_content(manifest_root)
        package_xml = convert_package_xml_to_catkin_package(package_xml_str)

        # Get requirements from package.xml
        package_requirements = package_xml_to_conda_requirements(package_xml, distro)

        # Add standard dependencies
        build_deps = ["ninja", "python", "setuptools", "git", "git-lfs", "cmake", "cpython"]
        if host_platform.is_unix:
            build_deps.extend(["patch", "make", "coreutils"])
        if host_platform.is_windows:
            build_deps.extend(["m2-patch"])
        if host_platform.is_osx:
            build_deps.extend(["tapi"])

        for dep in build_deps:
            package_requirements.build.append(ItemPackageDependency(name=dep))

        # Add compiler dependencies
        package_requirements.build.append(ItemPackageDependency("${{ compiler('c') }}"))
        package_requirements.build.append(ItemPackageDependency("${{ compiler('cxx') }}"))

        host_deps = ["python", "numpy", "pip", "pkg-config"]

        for dep in host_deps:
            package_requirements.host.append(ItemPackageDependency(name=dep))

        # Merge package requirements into the model requirements
        requirements = merge_requirements(generated_recipe.recipe.requirements, package_requirements)
        generated_recipe.recipe.requirements = requirements


        # Determine build platform
        build_platform = BuildPlatform.current()

        # Generate build script
        build_script_context = BuildScriptContext.load_from_template(package_xml, build_platform, manifest_root, distro)
        build_script_lines = build_script_context.render()

        generated_recipe.recipe.build.script =  Script(
            content=build_script_lines,
            env=backend_config.env,
        )

        debug_dir = backend_config.get_debug_dir()
        if debug_dir:
            recipe = generated_recipe.recipe.to_yaml()
            package = generated_recipe.recipe.package
            debug_file_path = debug_dir / f"{package.name.get_concrete()}-{package.version}-recipe.yaml"
            debug_file_path.parent.mkdir(parents=True, exist_ok=True)
            with open(debug_file_path, 'w') as debug_file:
                debug_file.write(recipe)

        # Test the build script before running to early out.
        # TODO: returned script.content list is not a list of strings, a container for that
        # so it cant be compared directly with the list yet
        # assert generated_recipe.recipe.build.script.content == build_script_lines, f"Script content {generated_recipe.recipe.build.script.content}, build script lines {build_script_lines}"
        return generated_recipe

    def extract_input_globs_from_build(self, config: ROSBackendConfig, editable: bool) -> List[str]:
        """Extract input globs for the build."""
        return get_build_input_globs(config, editable)



def merge_requirements(model_requirements: ConditionalRequirements, package_requirements: ConditionalRequirements) -> ConditionalRequirements:
    """Merge two sets of requirements."""
    merged = ConditionalRequirements()

    # The model requirements are the base, coming from the pixi manifest
    # We need to only add the names for non-existing dependencies
    def merge_unique_items(
            model: List[ItemPackageDependency],
            package: List[ItemPackageDependency],
    ) -> List[ItemPackageDependency]:
        """Merge unique items from source into target."""
        result = model

        for item in package:
            package_names = [i.concrete.package_name for i in model if i.concrete]

            if item.concrete is not None and item.concrete.package_name not in package_names:
                result.append(item)
            if str(item.template) not in [str(i.template) for i in model]:
                result.append(item)
        return result

    merged.host = merge_unique_items(model_requirements.host, package_requirements.host)
    merged.build = merge_unique_items(model_requirements.build, package_requirements.build)
    merged.run = merge_unique_items(model_requirements.run, package_requirements.run)

    # If the dependency is of type Source in one of the requirements, we need to set them to Source for all variants
    return merged


