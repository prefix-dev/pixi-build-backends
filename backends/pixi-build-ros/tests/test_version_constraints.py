from pathlib import Path
import pytest
import tempfile
from pixi_build_backend.types.platform import Platform
from pixi_build_backend.types.project_model import ProjectModelV1

from pixi_build_ros.ros_generator import ROSGenerator


def test_generate_recipe_with_versions(package_xmls: Path, test_data_dir: Path):
    """Test the generate_recipe function of ROSGenerator."""
    # Create a temporary directory to simulate the package directory
    with tempfile.TemporaryDirectory() as temp_dir:
        temp_path = Path(temp_dir)

        # Copy the test package.xml to the temp directory
        package_xml_source = package_xmls / "version_constraints.xml"
        package_xml_dest = temp_path / "package.xml"
        package_xml_dest.write_text(package_xml_source.read_text(encoding="utf-8"))

        # Create a minimal ProjectModelV1 instance
        model = ProjectModelV1()

        # Create config for ROS backend
        config = {
            "distro": "noetic",
            "noarch": False,
            "extra-package-mappings": [str(test_data_dir / "other_package_map.yaml")],
        }

        # Create host platform
        host_platform = Platform.current()

        # Create ROSGenerator instance
        generator = ROSGenerator()

        # Generate the recipe
        generated_recipe = generator.generate_recipe(
            model=model,
            config=config,
            manifest_path=str(temp_path),
            host_platform=host_platform,
        )

        # Verify the generated recipe has the expected requirements
        assert generated_recipe.recipe.package.name.get_concrete() == "ros-noetic-custom-ros"

        req_string = list(str(req) for req in generated_recipe.recipe.requirements.run)
        assert "ros-noetic-ros-package <2.0.0" in req_string
        assert "qt-main >=5.15.0,<5.16.0" in req_string


def test_wrong_version_constraints(package_xmls: Path, test_data_dir: Path):
    """Test the generate_recipe function of ROSGenerator."""
    # Create a temporary directory to simulate the package directory
    with tempfile.TemporaryDirectory() as temp_dir:
        temp_path = Path(temp_dir)

        # Copy the test package.xml to the temp directory
        package_xml_source = package_xmls / "version_constraints_wrong.xml"
        package_xml_dest = temp_path / "package.xml"
        package_xml_dest.write_text(package_xml_source.read_text(encoding="utf-8"))

        # Create a minimal ProjectModelV1 instance
        model = ProjectModelV1()

        # Create config for ROS backend
        config = {
            "distro": "noetic",
            "noarch": False,
            "extra-package-mappings": [str(test_data_dir / "other_package_map.yaml")],
        }

        # Create host platform
        host_platform = Platform.current()

        # Create ROSGenerator instance
        generator = ROSGenerator()

        with pytest.raises(ValueError) as excinfo:
            # Generate the recipe
            generator.generate_recipe(
                model=model,
                config=config,
                manifest_path=str(temp_path),
                host_platform=host_platform,
            )
        assert "Version specifier can only be used for one package" in excinfo.value.args[0]
