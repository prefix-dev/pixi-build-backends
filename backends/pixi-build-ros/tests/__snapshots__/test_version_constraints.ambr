# serializer version: 1
# name: test_generate_recipe_with_mutex_version
  list([
    'ros-noetic-catkin',
    'ros-noetic-ros-package <=2.0.0',
    'ros-noetic-ros-package-msgs <=2.0.0',
    'ros_package2 <2.0.0a0',
    'qt-main >=5.15.0,<5.16.0',
    'libgl-devel',
    'libopengl-devel',
    'xorg-libx11',
    'xorg-libxext',
    'tinyxml2 ==10.0.0',
    'python',
    'numpy',
    'pip',
    'pkg-config',
    'ros-distro-mutex 0.5.* noetic',
  ])
# ---
# name: test_generate_recipe_with_versions
  list([
    'ros-noetic-ros-package <=2.0.0',
    'ros-noetic-ros-package-msgs <=2.0.0',
    'ros_package2 <2.0.0a0',
    'qt-main >=5.15.0,<5.16.0',
    'tinyxml2 ==10.0.0',
  ])
# ---
